test/studies/adventOfCode/2022/day11/bradc/day11-par.chpl

```diff
@@ -0,0 +1,181 @@
+use IO, List, Barriers;
+
+config const numRounds = 20;
+
+// things that define a monkey
+class Monkey {
+  var id: int,                 // an integer ID
+      Items: [0..1] list(int, parSafe=true),  // two lists of its items
+      current = 0, next=1,     // which list is the current vs. next one
+      op: owned MathOp,            // the operator the monkey does
+      divisor: int,            // the divisor for its % operation check
+      targetMonkey: 2*int,     // the target monkeys it throws items to
+      numInspected: int;       // the count of how many inspections we've done
+}
+
+/*
+var m1 = new Monkey();
+var m2 = new Monkey();
+var m3 = new Monkey();
+var m4 = new Monkey();
+*/
+
+// An array of monkeys
+var Monkeys = readMonkeys();
+//writeln(Monkeys);
+
+
+// This tells whether a given monkey can proceed when it is out of items.
+// Initially, only monkey 0 can since nobody can throw items into its
+// list for this round, only for subsequent rounds.
+var canPause: sync int = 0;
+
+
+// Our main parallel simulation loop for the monkeys
+var bar = new Barrier(Monkeys.size);
+coforall monkey in Monkeys {
+  for r in 1..numRounds {
+    // Process any items that are in our list of current items
+    monkey.processItems(canPause);
+
+    // Make sure everyone is done so that we don't start changing our
+    // items list for the next round while monkeys are still adding
+    // things to it.
+    bar.barrier();
+
+    // Reset our items lists so that the list for the next round
+    // bcomes the new one
+    monkey.resetForNextRound(canPause);
```

Comment:
Can we take `canPause` by ref into monkey and not have to pass it like this in the loops? Would that be cleaner?

Reply:
It would be cleaner, but records and classes don't support `ref` fields yet, unfortunately.  Definitely a hotly desired feature by both users and our team.  If you see another way to thread it through the monkey, let me know.  Early drafts of the program just referred to it non-locally, but I felt weird about having the methods refer to globals... though they do in other cases (like `Monkeys`), so I'd be willing to switch it back if you preferred (or to pass in `Monkeys` if you liked that better.

Reply:
I think it's fine as is if we can't pass it in as a ref.

Reply:
@DanilaFe : Just to make doubly-sure:  You're OK with the inconsistency of passing in the global `canPause` but not the global `Monkeys` (vs. passing in neither or both?)

Reply:
Yes, I am okay with it.